File: baekjoon/StringQuestion/034-5582-공통부분문자열.java
import java.io.BufferedReader;
import java.io.InputStreamReader;


public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        char[] str1 = br.readLine().replace("\n", "").toCharArray();
        char[] str2 = br.readLine().replace("\n", "").toCharArray();
        int[][] dp = new int[str1.length][str2.length];

        int max = 0;
        for (int i = 0; i < str1.length; i++) {
            for (int j = 0; j < str2.length; j++) {
                if (str1[i] == str2[j]) {
                    if (i > 0 && j > 0 && dp[i-1][j-1] > 0) {
                        dp[i][j] += dp[i-1][j-1] + 1;
                        if (dp[i][j] > max) max = dp[i][j];
                    }
                    else dp[i][j] = 1;
                }
            }
        }

        System.out.print(max);
    }
}

File: baekjoon/StringQuestion/034-5582-공통부분문자열.py
import sys 


input = sys.stdin.readline
str1 = input().rstrip()
str2 = input().rstrip()

dp = [[0 for _ in range(len(str1))] for _ in range(len(str2))]
maxleng = 0
for i in range(len(str2)):
    for j in range(len(str1)):
        if str2[i] == str1[j]:
            if i > 0 and j > 0 and dp[i-1][j-1] > 0:
                dp[i][j] += dp[i-1][j-1] + 1
                if dp[i][j] > maxleng: maxleng = dp[i][j]
            else:
                dp[i][j] = 1

for i in range(len(str2)):
    print(dp[i])

print(maxleng)